### PR TITLE
[wip][pt1][tensor] BlobGetMutableTensor returns Tensor

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#14136 [wip][pt1][tensor] BlobGetMutableTensor returns Tensor**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D12934074/)

Since now Tensor is a shared_ptr, it doesn't make sense to have Tensor* around anymore,
this diff is changing Tensor* to Tensor in the interface.
We'll store pointer to intrusive_ptr to TensorImpl in Tensor and create Tensor object on  the fly
when returning from BlobGetMutableTensor.
Another motivation is we want Blob to store intrusive_ptr of TensorImpl because IValue can store intrusive ptr. Without this change, IValue will have to store Blob which is an extra level of indirection.

This diff should be rebased on the diffs that removes Resize/ResizeLike+mutable_data

To remove Tensor*, we'll do following
```
auto* Y = Ouptut(0);
Y->mutable_data...
```
-->
```
auto Y = Output(0);
Y.mutable_data...
```

But to run clangr codemod, we'll keep both APIs in different names, e.g. `Output` and `XOutput`, and do the refactor and then delete the old method and rename the new method into the old one.
For example for `Output`, we'll first codemod the callsites from `Output` to `XOutput`, then delete the old `Output` and rename `XOutput` to `Output` in the end.

Differential Revision: [D12934074](https://our.internmc.facebook.com/intern/diff/D12934074/)

### CODE REVIEW

File: c10/util/typeid.cpp
Comment (ezyang) [inline c10/util/typeid.cpp]: BUG FARM!!!!

File: caffe2/core/blob.h
Comment (ezyang) [inline caffe2/core/blob.h]: What does the X mean?

File: caffe2/core/tensor.h
Comment (ezyang) [inline caffe2/core/tensor.h]: Ahem. If you're going to copy implementations from `at::Tensor`, we should do it right. You should `std::move()` the pointer here, or you're going to do an unnecessary extra copy.

File: caffe2/core/tensor.h
Comment (ezyang) [inline caffe2/core/tensor.h]: This method diverges from what is in at::Tensor. Let's keep them lined up.

File: caffe2/core/blob.h
Comment (jerryzh168) [reply caffe2/core/blob.h]: nothing, just a temporary name for codemod, we'll change remove the other BlobIsTensorType and change this to BlobIsTensorType after the codemod is done.

**Comment by jerryzh168** [comment]
@ezyang yeah, in terms of Blob->Tensor it's the same, this diff is useful for changing Blob to IValue, because right now IValue have to store a Blob -> Tensor and with this diff IValue can store IValue -> intrusive_ptr to TensorImpl. cc @smessmer for more details.

**Comment by ezyang** [comment]
So, is the point to change the signatures of the functions in question? That seems fine, but you don't need to actually change the internal implementation to change the signatures, no?

**Comment by jerryzh168** [comment]
We want to change both signature and implementation, we do want to have Blob store TensorImplPtr to facilitate the use of IValue in Operators.

**Comment by dzhulgakov** [comment]
Hm, I agree with Ed - the change of what you store in Blob is independent from what the API of BlobGetMutableTensor is.

You can store the caffe2::Tensor and still extract intrusivePtr from it if necessary

**Comment by jerryzh168** [comment]
@dzhulgakov I think we want to change both API and underlying implementation. not just API.

**Review by smessmer — APPROVED** [review]
This diff doesn't change anything in the indirections, before it was `Blob` stores `Tensor*`, `Tensor` stores `intrusive_ptr<TensorImpl>`. After this diff, it is `Blob` stores `intrusive_ptr<TensorImpl>*`, both are two indirections. This is, however, not the goal of this diff, and will be fixed once we move to `IValue`, where we would directly store `TensorImpl*`.

The goal of this diff is to unblock the `Blob` -> `IValue` merge. When `IValue` stores Tensors, it already does the right thing and stores them as `TensorImpl*`. I don't want to add the ability to store `Tensor*` to `IValue`. That means, however, that if we had an accessor function returning `Tensor*`, it would be a pointer to a temporary object and stuff would break. So we need all accessor functions to return `Tensor` by value. And this diff does that for Blob accessors, so we can then replace `Blob` with `IValue`. That's the API part.

Another thing is that this diff does is change the implementation. We want to be able to easily convert `Blob` into `IValue` and the other way round. While that is possible when `Blob` stores `Tensor*`, it's easier if they have the same way of storing Tensors. If there's a strong argument against changing the implementation, we don't have to, but I don't see why not.

**Comment by jerryzh168** [comment]
My plan is include the codemod in this diff as well, since we'll be able to remove the temporary function that starts with X after the codemod, or should I land this first?

File: caffe2/core/tensor.cc
Comment (jerryzh168) [reply caffe2/core/tensor.cc]: sure

**Comment by smessmer** [comment]
Land this first, it's easier to review if the codemod is separate.

**Comment by ezyang** [comment]
@smessmer The API change seems fine. But I'm still confused about the implementation change.

> Another thing is that this diff does is change the implementation. We want to be able to easily convert Blob into IValue and the other way round. While that is possible when Blob stores Tensor*, it's easier if they have the same way of storing Tensors. If there's a strong argument against changing the implementation, we don't have to, but I don't see why not.

But even after this patch they don't have the same way of storing tensors. ivalue stores:

```
    c10::intrusive_ptr_target* as_intrusive_ptr;
```

That's not what this patch is changing the implementation to.

**Comment by jerryzh168** [comment]
just talked with @smessmer, we'll keep this diff API change only, and we can postpone the implementation change until we are replacing Blob with IValue.

**Pull request closed without merging**
